numpy/lib/utils.py

```diff
@@ -193,7 +193,43 @@ def deprecate(*args, **kwargs):
     else:
         return _Deprecate(*args, **kwargs)
 
-deprecate_with_doc = lambda msg: _Deprecate(message=msg)
+
+def deprecate_with_doc(msg):
+    """
+    Deprecates a function and includes the deprecation in its docstring.
+    
+    This function is used as a decorator. It returns an object that can be 
+    used to issue a DeprecationWarning, by passing the to-be decorated 
+    function as argument, this adds warning to the to-be decorated function's 
+    docstring and returns the new function object.
+    
+    See Also
+    --------
+    deprecate : Decorate a function such that it issues a `DeprecationWarning` 
+    
+    Parameters
+    ----------
+    msg : str
+        Additional explanation of the deprecation. Displayed in the 
+        docstring after the warning.
+
+    Returns
+    -------
+    obj : object
+
+    Examples
+    --------
+    Note that ``olduint`` returns a value after raising DeprecationWarning 
+    with `msg`:
+
+    >>> oldobj = np.deprecate_with_doc("Use np.int_ instead.")
+    >>> olduint = oldobj(np.uint)
+    >>> olduint(6)
+    DeprecationWarning: `uint64` is deprecated! #may vary 
+        Use np.int_ instead.		        	
+    6
```

Comment:
This example is flawed since `np.uint` varies acrosss platforms. Maybe something like (please verify it outputs the correct text or the refguide tests will continue to fail):
```
Examples
--------

>>> @np.deprecate_with_doc("Use newfunc instead")
... def oldfunc(a):
...     return a
>>> 
>>> oldfunc(3)
DeprecationWarning: oldfunc is deprecated!
    Use newfunc instead
3
```

Reply:
> This example is flawed since `np.uint` varies acrosss platforms. Maybe something like (please verify it outputs the correct text or the refguide tests will continue to fail):
> 
> ```
> Examples
> --------
> 
> >>> @np.deprecate_with_doc("Use newfunc instead")
> ... def oldfunc(a):
> ...     return a
> >>> 
> >>> oldfunc(3)
> DeprecationWarning: oldfunc is deprecated!
>     Use newfunc instead
> 3
> ```

I have made the required changes however I believe the `DeprecationWarning` was not being shown in any case which was causing the tests to fail. I have added a `.__doc__`  statement in the example as an alternative for the same.

Reply:
The refguide tests seem to be failing to due to the newline character `\n` not being interpreted properly. The output is `oldfunc is deprecated!\nUse newfunc instead` instead of the expected output. The same works fine otherwise.

Reply:
It's not the newline that's causing issues here, it's the indentation of.
The 4 spaces preceding `Use newfunc instead` shouldn't be there.

Reply:
I got rid of the the 4 spaces but the tests still fail because of `oldfunc is deprecated!\nUse newfunc instead`.